File: frontend/src/lib/pays.ts
/**
 * Liste des pays proposés à l'inscription et dans l'administration (valeur stockée : le nom
 * français, tel qu'affiché dans les listes admin). L'indicatif sert d'aide de saisie du
 * téléphone — le numéro n'est pas forcément un compte mobile money.
 */
import type { GroupeOptions, OptionSelect } from '@/components/partages/select/select'

export interface Pays {
  code: string
  nom: string
  indicatif: string
}

export const PAYS: readonly Pays[] = [
  { code: 'CI', nom: "Côte d'Ivoire", indicatif: '+225' },
  { code: 'SN', nom: 'Sénégal', indicatif: '+221' },
  { code: 'ML', nom: 'Mali', indicatif: '+223' },
  { code: 'BF', nom: 'Burkina Faso', indicatif: '+226' },
  { code: 'TG', nom: 'Togo', indicatif: '+228' },
  { code: 'BJ', nom: 'Bénin', indicatif: '+229' },
  { code: 'NE', nom: 'Niger', indicatif: '+227' },
  { code: 'GN', nom: 'Guinée', indicatif: '+224' },
  { code: 'GW', nom: 'Guinée-Bissau', indicatif: '+245' },
  { code: 'MR', nom: 'Mauritanie', indicatif: '+222' },
  { code: 'GM', nom: 'Gambie', indicatif: '+220' },
  { code: 'SL', nom: 'Sierra Leone', indicatif: '+232' },
  { code: 'LR', nom: 'Liberia', indicatif: '+231' },
  { code: 'GH', nom: 'Ghana', indicatif: '+233' },
  { code: 'NG', nom: 'Nigeria', indicatif: '+234' },
  { code: 'CM', nom: 'Cameroun', indicatif: '+237' },
  { code: 'GA', nom: 'Gabon', indicatif: '+241' },
  { code: 'CG', nom: 'Congo', indicatif: '+242' },
  { code: 'CD', nom: 'RD Congo', indicatif: '+243' },
  { code: 'CF', nom: 'Centrafrique', indicatif: '+236' },
  { code: 'TD', nom: 'Tchad', indicatif: '+235' },
  { code: 'GQ', nom: 'Guinée équatoriale', indicatif: '+240' },
  { code: 'MA', nom: 'Maroc', indicatif: '+212' },
  { code: 'DZ', nom: 'Algérie', indicatif: '+213' },
  { code: 'TN', nom: 'Tunisie', indicatif: '+216' },
  { code: 'EG', nom: 'Égypte', indicatif: '+20' },
  { code: 'KE', nom: 'Kenya', indicatif: '+254' },
  { code: 'ET', nom: 'Éthiopie', indicatif: '+251' },
  { code: 'TZ', nom: 'Tanzanie', indicatif: '+255' },
  { code: 'UG', nom: 'Ouganda', indicatif: '+256' },
  { code: 'RW', nom: 'Rwanda', indicatif: '+250' },
  { code: 'BI', nom: 'Burundi', indicatif: '+257' },
  { code: 'MG', nom: 'Madagascar', indicatif: '+261' },
  { code: 'MU', nom: 'Maurice', indicatif: '+230' },
  { code: 'ZA', nom: 'Afrique du Sud', indicatif: '+27' },
  { code: 'AO', nom: 'Angola', indicatif: '+244' },
  { code: 'MZ', nom: 'Mozambique', indicatif: '+258' },
  { code: 'ZM', nom: 'Zambie', indicatif: '+260' },
  { code: 'FR', nom: 'France', indicatif: '+33' },
  { code: 'BE', nom: 'Belgique', indicatif: '+32' },
  { code: 'CH', nom: 'Suisse', indicatif: '+41' },
  { code: 'GB', nom: 'Royaume-Uni', indicatif: '+44' },
  { code: 'DE', nom: 'Allemagne', indicatif: '+49' },
  { code: 'ES', nom: 'Espagne', indicatif: '+34' },
  { code: 'IT', nom: 'Italie', indicatif: '+39' },
  { code: 'PT', nom: 'Portugal', indicatif: '+351' },
  { code: 'CA', nom: 'Canada', indicatif: '+1' },
  { code: 'US', nom: 'États-Unis', indicatif: '+1' },
  { code: 'BR', nom: 'Brésil', indicatif: '+55' },
  { code: 'AE', nom: 'Émirats arabes unis', indicatif: '+971' },
  { code: 'TR', nom: 'Turquie', indicatif: '+90' },
  { code: 'IN', nom: 'Inde', indicatif: '+91' },
  { code: 'CN', nom: 'Chine', indicatif: '+86' },
]

/** Pays proposés en tête de liste (marché principal). */
const FREQUENTS = ['CI', 'SN', 'ML', 'BF', 'TG', 'BJ', 'CM', 'GN']

const parNom = (a: Pays, b: Pays) => a.nom.localeCompare(b.nom, 'fr')

/** Groupes `<optgroup>` : pays fréquents puis tous les pays par ordre alphabétique. */
export function optionsPaysGroupees(): GroupeOptions[] {
  const frequents = FREQUENTS.map((code) => PAYS.find((p) => p.code === code)).filter((p): p is Pays => p !== undefined)
  const tous = [...PAYS].sort(parNom)
  const enOption = (p: Pays): OptionSelect => ({ valeur: p.nom, libelle: `${p.nom} (${p.indicatif})` })
  return [
    { libelle: 'Pays fréquents', options: frequents.map(enOption) },
    { libelle: 'Tous les pays', options: tous.map(enOption) },
  ]
}

/** Retrouve un pays par nom (valeur stockée) ou par code ISO. */
export function trouverPays(valeur: string | null | undefined): Pays | undefined {
  if (!valeur) return undefined
  const v = valeur.trim().toLowerCase()
  return PAYS.find((p) => p.nom.toLowerCase() === v || p.code.toLowerCase() === v)
}

/** Indicatif téléphonique du pays (ou chaîne vide si inconnu). */
export function indicatifPays(valeur: string | null | undefined): string {
  return trouverPays(valeur)?.indicatif ?? ''
}

File: mobile/lib/noyau/pays.dart
/// Pays proposés à l'inscription et au profil — transcription de
/// `frontend/src/lib/pays.ts`. La valeur STOCKÉE est le nom français, exactement
/// comme sur le web : les deux clients écrivent la même chose en base.
///
/// L'indicatif ne sert qu'à préremplir la saisie du téléphone ; le numéro n'est
/// pas forcément un compte Mobile Money.
class Pays {
  const Pays(this.code, this.nom, this.indicatif);
  final String code;
  final String nom;
  final String indicatif;
}

const List<Pays> paysConnus = [
  Pays('CI', "Côte d'Ivoire", '+225'),
  Pays('SN', 'Sénégal', '+221'),
  Pays('ML', 'Mali', '+223'),
  Pays('BF', 'Burkina Faso', '+226'),
  Pays('TG', 'Togo', '+228'),
  Pays('BJ', 'Bénin', '+229'),
  Pays('NE', 'Niger', '+227'),
  Pays('GN', 'Guinée', '+224'),
  Pays('GW', 'Guinée-Bissau', '+245'),
  Pays('MR', 'Mauritanie', '+222'),
  Pays('GM', 'Gambie', '+220'),
  Pays('SL', 'Sierra Leone', '+232'),
  Pays('LR', 'Liberia', '+231'),
  Pays('GH', 'Ghana', '+233'),
  Pays('NG', 'Nigeria', '+234'),
  Pays('CM', 'Cameroun', '+237'),
  Pays('GA', 'Gabon', '+241'),
  Pays('CG', 'Congo', '+242'),
  Pays('CD', 'RD Congo', '+243'),
  Pays('CF', 'Centrafrique', '+236'),
  Pays('TD', 'Tchad', '+235'),
  Pays('GQ', 'Guinée équatoriale', '+240'),
  Pays('MA', 'Maroc', '+212'),
  Pays('DZ', 'Algérie', '+213'),
  Pays('TN', 'Tunisie', '+216'),
  Pays('EG', 'Égypte', '+20'),
  Pays('KE', 'Kenya', '+254'),
  Pays('ET', 'Éthiopie', '+251'),
  Pays('TZ', 'Tanzanie', '+255'),
  Pays('UG', 'Ouganda', '+256'),
  Pays('RW', 'Rwanda', '+250'),
  Pays('BI', 'Burundi', '+257'),
  Pays('MG', 'Madagascar', '+261'),
  Pays('MU', 'Maurice', '+230'),
  Pays('ZA', 'Afrique du Sud', '+27'),
  Pays('AO', 'Angola', '+244'),
  Pays('MZ', 'Mozambique', '+258'),
  Pays('ZM', 'Zambie', '+260'),
  Pays('FR', 'France', '+33'),
  Pays('BE', 'Belgique', '+32'),
  Pays('CH', 'Suisse', '+41'),
  Pays('GB', 'Royaume-Uni', '+44'),
  Pays('DE', 'Allemagne', '+49'),
  Pays('ES', 'Espagne', '+34'),
  Pays('IT', 'Italie', '+39'),
  Pays('PT', 'Portugal', '+351'),
  Pays('CA', 'Canada', '+1'),
  Pays('US', 'États-Unis', '+1'),
  Pays('BR', 'Brésil', '+55'),
  Pays('AE', 'Émirats arabes unis', '+971'),
  Pays('TR', 'Turquie', '+90'),
  Pays('IN', 'Inde', '+91'),
  Pays('CN', 'Chine', '+86'),
];

/// Pays proposés en tête de liste (marché principal).
const List<String> _frequents = ['CI', 'SN', 'ML', 'BF', 'TG', 'BJ', 'CM', 'GN'];

List<Pays> get paysFrequents => _frequents
    .map((code) => paysConnus.where((p) => p.code == code).firstOrNull)
    .whereType<Pays>()
    .toList(growable: false);

List<Pays> get paysParNom {
  final tous = [...paysConnus]..sort((a, b) => a.nom.compareTo(b.nom));
  return tous;
}

/// Retrouve un pays par nom (valeur stockée) ou par code ISO.
Pays? trouverPays(String? valeur) {
  if (valeur == null || valeur.trim().isEmpty) return null;
  final v = valeur.trim().toLowerCase();
  return paysConnus
      .where((p) => p.nom.toLowerCase() == v || p.code.toLowerCase() == v)
      .firstOrNull;
}

/// Indicatif téléphonique du pays, ou chaîne vide s'il est inconnu.
String indicatifPays(String? valeur) => trouverPays(valeur)?.indicatif ?? '';

extension _Premier<T> on Iterable<T> {
  T? get firstOrNull {
    final it = iterator;
    return it.moveNext() ? it.current : null;
  }
}
